src/core_functions/scalar/list/list_reduce.cpp

```diff
@@ -0,0 +1,217 @@
+#include "duckdb/core_functions/scalar/list_functions.hpp"
+#include "duckdb/core_functions/lambda_functions.hpp"
+#include "duckdb/planner/expression/bound_cast_expression.hpp"
+#include "duckdb/planner/expression/bound_function_expression.hpp"
+
+namespace duckdb {
+
+struct ReduceExecuteInfo {
+	ReduceExecuteInfo(LambdaFunctions::LambdaInfo &info, ClientContext &context) : left_slice(*info.child_vector) {
+		SelectionVector left_vector(info.row_count);
+		active_rows.resize(info.row_count);
+
+		idx_t old_count = 0;
+		idx_t new_count = 0;
+
+		// Initialize the left vector from list entries and the active rows
+		auto it = active_rows.begin();
+		while (it != active_rows.end()) {
+			auto list_column_format_index = info.list_column_format.sel->get_index(old_count);
+			if (info.list_column_format.validity.RowIsValid(list_column_format_index)) {
+				if (info.list_entries[list_column_format_index].length == 0) {
+					throw ParameterNotAllowedException("Cannot perform list_reduce on an empty input list");
+				}
+				left_vector.set_index(new_count, info.list_entries[list_column_format_index].offset);
+				active_rows[new_count] = old_count;
+				new_count++;
+				it++;
+			} else {
+				// Remove the invalid rows
+				info.result_validity->SetInvalid(old_count);
+				active_rows.erase(it);
+			}
+			old_count++;
+		}
+
+		left_slice.Slice(left_vector, new_count);
+
+		if (info.has_index) {
+			input_types.push_back(LogicalType::BIGINT);
+		}
+		input_types.push_back(left_slice.GetType());
+		input_types.push_back(left_slice.GetType());
+		for (auto &entry : info.column_infos) {
+			input_types.push_back(entry.vector.get().GetType());
+		}
+
+		expr_executor = make_uniq<ExpressionExecutor>(context, *info.lambda_expr);
+	};
+
+	vector<idx_t> active_rows;
```

Comment:
Can't this be a `SelectionVector`?

Reply:
Ah, good point. I think this can even be the same selection vector as the one for the non-lambda parameter inputs. @maiadegraaf?

Reply:
Yes exactly, I think that's an easy switch